### PR TITLE
There is no need to split service key repeatedly

### PR DESCRIPTION
**Pull request by FengyunPan**
**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -319,6 +319,8 @@ func (e *EndpointController) syncService(key string) error {
 
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Syncing service %q endpoints, but couldn't understand the key: %v", key, err))
```
Comment (bowei) [inline pkg/controller/endpoint/endpoints_controller.go]: Can you change the error message to:

fmt.Errorf("syncService(%q), invalid key: %v", key, err)

File: pkg/controller/endpoint/endpoints_controller.go [hunk 2]
```diff
@@ -319,6 +319,8 @@ func (e *EndpointController) syncService(key string) error {
...
+		// Don't retry, as the key isn't going to magically become understandable.
```
Comment (MrHohn) [inline pkg/controller/endpoint/endpoints_controller.go]: This comment says don't retry, but you still return error below?

File: pkg/controller/endpoint/endpoints_controller.go (hunk 1 above)
Comment (FengyunPan) [inline pkg/controller/endpoint/endpoints_controller.go]: Thank you, done.

File: pkg/controller/endpoint/endpoints_controller.go (hunk 2 above)
Comment (FengyunPan) [inline pkg/controller/endpoint/endpoints_controller.go]: Oops, I update it soon.

**Review by MrHohn — APPROVED** [review]
Thanks, LGTM

**Comment by bowei** [comment]
Actually, taking a closer look at this, you should fix this by moving the parsing of the key into `processNextWorkItem`. If the key cannot be parsed when the item is popped out of the queue, log and error and remove the item from the queue. This will make it very explicit that the item is being dropped from the queue:

```
func (e *EndpointController) processNextWorkItem() bool {
	eKey, quit := e.queue.Get()
	if quit {
		return false
	}
	defer e.queue.Done(eKey)
       	namespace, name, err := cache.SplitMetaNamespaceKey(ekey.(string))
	if err != nil {
              glog.Errorf("Invalid key in work queue (%q): %v", ekey.(string), err)
              return true
	}
	err := e.syncService(namespace, name)
	e.handleErr(err, eKey)

	return true
}

func (e *EndpointController) syncService(namespace, name string) error {
...
```

**Comment by FengyunPan** [comment]
@bowei Nice comment, thank you very much, I update it soon.

**Comment by FengyunPan** [comment]
/assign @derekwaynecarr
Can you approve this pr?

**Comment by FengyunPan** [comment]
/assign @deads2k
Can you approve this pr?

File: pkg/controller/endpoint/endpoints_controller.go [hunk 3]
```diff
@@ -311,29 +317,19 @@ func (e *EndpointController) handleErr(err error, key interface{}) {
 	utilruntime.HandleError(err)
 }
 
-func (e *EndpointController) syncService(key string) error {
```
Comment (deads2k) [inline pkg/controller/endpoint/endpoints_controller.go]: Having the key here is consistent with all the other controllers.  How about just skipping the second split in this method and us ethe already split values.

Comment (FengyunPan) [inline pkg/controller/endpoint/endpoints_controller.go]: Ok, done.

**Pull request merged**